channeld/channeld.c

```diff
@@ -2199,40 +2586,1304 @@ static void handle_unexpected_reestablish(struct peer *peer, const u8 *msg)
+	txsig_tlvs->funding_outpoint_sig = tal_dup_arr(tmpctx, u8, der,
+						       der_len, 0);
+
+	ws = psbt_to_witness_stacks(tmpctx, signed_psbt,
+				    TX_INITIATOR, splice_funding_index);
+	sigout_msg = towire_tx_signatures(tmpctx, &peer->channel_id,
+					  &signed_psbt_txid, ws, txsig_tlvs);
+
+	if (do_i_sign_first(peer, signed_psbt, TX_INITIATOR)) {
+		status_debug("Splice initiator: we sign first");
+		outmsg = towire_channeld_update_inflight(NULL, signed_psbt);
+		wire_sync_write(MASTER_FD, take(outmsg));
+		peer_write(peer->pps, take(sigout_msg));
+	}
+
+	msg = peer_read(tmpctx, peer->pps);
+	if (handle_peer_error(peer->pps, &peer->channel_id, msg))
+		return;
+
+	if (fromwire_peektype(msg) != WIRE_TX_SIGNATURES)
+		peer_failed_warn(peer->pps, &peer->channel_id,
+				"Splicing got incorrect message from peer: %s "
+				"(should be WIRE_TX_SIGNATURES)",
+				peer_wire_name(fromwire_peektype(msg)));
+
+	if (!fromwire_tx_signatures(tmpctx, msg, &cid, &txid,
+				    cast_const3(struct witness_stack ***, &ws),
+				    &txsig_tlvs))
+		peer_failed_warn(peer->pps, &peer->channel_id,
+			    "Splicing bad tx_signatures %s",
+			    tal_hex(msg, msg));
+
+	their_sig.sighash_type = SIGHASH_ALL;
+
+	if (!signature_from_der(txsig_tlvs->funding_outpoint_sig,
+			       tal_count(txsig_tlvs->funding_outpoint_sig),
+			       &their_sig)) {
+
+		peer_failed_warn(peer->pps, &peer->channel_id,
+				 "Splicing bad tx_signatures %s",
+				 tal_hex(msg, msg));
+	}
+
+	their_pubkey = &peer->channel->funding_pubkey[REMOTE];
+
+	/* Set the commit_sig on the commitment tx psbt */
+	if (!psbt_input_set_signature(signed_psbt,
+				      splice_funding_index,
+				      their_pubkey,
+				      &their_sig)) {
+
+		status_failed(STATUS_FAIL_INTERNAL_ERROR,
+			      "Unable to set signature internally "
+			      "funding_index: %d "
+			      "pubkey: %s "
+			      "signature: %s "
+			      "psbt: %s",
+			      splice_funding_index,
+			      type_to_string(tmpctx, struct pubkey, their_pubkey),
+			      type_to_string(tmpctx, struct bitcoin_signature, &their_sig),
+			      type_to_string(tmpctx, struct wally_psbt, signed_psbt));
+	}
+
+	psbt_input_set_witscript(signed_psbt,
+				 splice_funding_index,
+				 wit_script);
+
+	final_tx = bitcoin_tx_with_psbt(tmpctx, signed_psbt);
+
+	wit_stack = bitcoin_witness_2of2(signed_psbt, &splice_sig, &their_sig,
+					 &peer->channel->funding_pubkey[LOCAL],
+					 their_pubkey);
+
+	bitcoin_tx_input_set_witness(final_tx, splice_funding_index, wit_stack);
+
+	/* DTODO: validate our peer's signatures are correct
+	 * see closingd.c receive_offer close_tx / check_tx_sig */
+
+	outmsg = towire_channeld_update_inflight(NULL, signed_psbt);
+	wire_sync_write(MASTER_FD, take(outmsg));
+
+	if (!do_i_sign_first(peer, signed_psbt, TX_INITIATOR)) {
+		status_debug("Splice accepter: we sign first");
+		peer_write(peer->pps, take(sigout_msg));
+	}
+
+	/* Sending of TX_SIGNATURE implies the end of stfu mode */
+	end_stfu_mode(peer);
```

Comment:
should we do it immediately after we send above then?

Reply:
might make sense to wait until we've called `reset_splice` before ending `stfu mode`.

Reply:
> should we do it immediately after we send above then?


Calling reset_splice after sending here seems dangerous. If we sign first -- then we're still waiting for the other person's signature.

I thiiiiink it would work today but this means we're ending splicing before receiving the final `WIRE_TX_SIGNATURES` which just feels vaguely odd. What if some code that receives `WIRE_TX_SIGNATURES` needs to access something about the splicing state that we already destroyed?

Reply:
> might make sense to wait until we've called `reset_splice` before ending `stfu mode`.

I think they're functionally the same, sure 👍